Monika After Story/game/script-topics.rpy

```diff
@@ -321,6 +321,81 @@ init 11 python:
 #        monika_random_topics=list(all_random_topics)
 
 
+default persistent._mas_player_bookmarked = {}
+# dict to store bookmarked events
+default persistent._mas_player_derandomed = {}
+# dict to store player derandomed events
+default persistent.flagged_monikatopic = None
+# var set when we flag a topic for derandom
+
+init python:
+    def mas_derandom_topic(ev_label=None):
+        """
+        Function for the derandom hotkey, 'x'
+
+        IN:
+            ev_label - label of the event we want to derandom.
+                (Optional, defaults to persistent.current_monikatopic)
+        """
+        if ev_label is None:
+            ev_label = persistent.current_monikatopic
+
+        ev = mas_getEV(ev_label)
+
+        if (
+            ev is not None
+            and ev.random
+            and ev.eventlabel.startswith("monika_")
+            and ev.prompt != ev.eventlabel
+        ):
+            if "mas_topic_derandom" not in persistent.event_list:
+                persistent.flagged_monikatopic = ev.eventlabel
+                pushEvent('mas_topic_derandom',True)
+                renpy.notify(__("Topic flagged for removal."))
+            else:
+                persistent.event_list.remove("mas_topic_derandom")
+                renpy.notify(__("Topic flag removed."))
+
+    def mas_bookmark_topic(ev_label=None):
+        """
+        Function for the bookmark hotkey, 'b'
+
+        IN:
+            ev_label - label of the event we want to bookmark.
+                (Optional, defaults to persistent.current_monikatopic)
+        """
+        if ev_label is None:
+            ev_label = persistent.current_monikatopic
+
+        ev = mas_getEV(ev_label)
+
+        if (
+            mas_isMoniNormal(higher=True)
+            and ev is not None
+            and ev.eventlabel.startswith("monika_")
+            and ev.prompt != ev.eventlabel
+        ):
+            if ev.eventlabel not in persistent._mas_player_bookmarked:
+                persistent._mas_player_bookmarked[ev.eventlabel] = ev
+                renpy.notify(__("Topic bookmarked."))
+            else:
+                persistent._mas_player_bookmarked.pop(ev.eventlabel)
+                renpy.notify(__("Bookmark removed."))
+
+    def mas_hasBookmarks():
+        """
+        Checks to see if we have bookmarks to show
+        """
+        if mas_isMoniNormal(higher=True):
+            bookmarkslist = [
+                (renpy.substitute(ev.prompt), ev_label, False, False)
+                for ev_label, ev in persistent._mas_player_bookmarked.iteritems()
+                if ev.unlocked and ev.checkAffection(mas_curr_affection)
+            ]
+            if len(bookmarkslist) > 0:
+                return True
```

Comment:
this can be way simpler:
```python
if mas_isMoniUpset(lower=True):
    return False

for ev in persistent._mas_player_bookmakred.itervalues():
    if ev.unlocked and ev.checkAffection(mas_curr_affection):
        return True
return False
```